accord-core/src/test/java/accord/impl/basic/DelayedCommandStores.java

```diff
@@ -18,26 +18,97 @@
     }
 
-    public static CommandStores.Factory factory(PendingQueue pending, RandomSource random)
+    public static class DelayedCommandStore extends InMemoryCommandStore
     {
-        SimulatedDelayedExecutorService executorService = new SimulatedDelayedExecutorService(pending, random);
-        return (time, agent, store, shardDistributor, progressLogFactory) -> new DelayedCommandStores(time, agent, store, shardDistributor, progressLogFactory, executorService);
+        private final SimulatedDelayedExecutorService executor;
+        private TaskExecutorService.Task<?> previous = null;
+
+        public DelayedCommandStore(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpochHolder rangesForEpochHolder, SimulatedDelayedExecutorService executor)
+        {
+            super(id, time, agent, store, progressLogFactory, rangesForEpochHolder);
+            this.executor = executor;
+        }
+
+        private static CommandStore.Factory factory(SimulatedDelayedExecutorService executor)
+        {
+            return (id, time, agent, store, progressLogFactory, rangesForEpoch) -> new DelayedCommandStore(id, time, agent, store, progressLogFactory, rangesForEpoch, executor);
+        }
+
+        @Override
+        public boolean inStore()
+        {
+            return CommandStore.Unsafe.maybeCurrent() == this;
+        }
+
+        @Override
+        public AsyncChain<Void> execute(PreLoadContext context, Consumer<? super SafeCommandStore> consumer)
+        {
+            return submit(context, i -> { consumer.accept(i); return null; });
+        }
+
+        @Override
+        public <T> AsyncChain<T> submit(PreLoadContext context, Function<? super SafeCommandStore, T> function)
+        {
+            return submit(() -> executeInContext(this, context, function));
+        }
+
+        @Override
+        public <T> AsyncChain<T> submit(Callable<T> fn)
+        {
+            TaskExecutorService.Task<T> task = new TaskExecutorService.Task<>(() -> Unsafe.runWith(this, fn));
+            task.addCallback(agent()); // used to track unexpected exceptions and notify simulations
+            if (previous == null || previous.isDone())
+            {
+                executor.execute(task);
+            }
+            else
+            {
+                // SimulatedDelayedExecutorService can interleave tasks and is global; this violates a requirement for
+                // CommandStore; single threaded with ordered execution!  To simulate this behavior, add the callback
+                // to enqueue once the "previous" task completes.
+                long nowMilis = executor.nowMillis();
+                previous.addCallback((i1, i2) -> executor.executeWithObservedDelay(task, executor.nowMillis() - nowMilis, TimeUnit.MILLISECONDS));
```

Comment:
a `FIFOQueue` would work in this case, let me test this out and get back to you

Reply:
pushed this change.  I think the code is more complex, but if you are in a debugger its easier to see when the queue is growing

Reply:
I think the code could be simplified by moving the callback logic to a member method that is _always_ registered as a callback to the currently executing task, e.g.

```
        private <T> void runNextTask()
        {
            Pending next = queue.peek();
            if (next == null)
                return;

            next.task.addCallback(agent()); // used to track unexpected exceptions and notify simulations
            next.task.addCallback(this::afterExecution);
            executor.executeWithObservedDelay(next.task, executor.nowMillis() - next.nowMillis, TimeUnit.MILLISECONDS);
        }

        private void afterExecution()
        {
            queue.poll();
            runNextTask();
        }

        @Override
        public <T> AsyncChain<T> submit(Callable<T> fn)
        {
            Task<T> task = new Task<>(() -> Unsafe.runWith(this, fn));
            boolean wasEmpty = queue.isEmpty();
            queue.add(new Pending(task, executor.nowMillis()));
            if (wasEmpty)
                runNextTask();
            return task;
        }
```

I will be honest, though, I don't quite follow the nowMillis delay logic. Could you explain your reasoning?

Reply:
> I will be honest, though, I don't quite follow the nowMillis delay logic. Could you explain your reasoning?

I just deleted my long post explaining my reasoning as after finishing it highlights that we *shouldn't* account for the queue delay...

The main reason was to avoid compounding the jitters (if you schedule task1, then task2, then task3, task3's jitter is the task1.jitter + task2.jitter + task3.jitter), but due to the execution forced to be in-order this has to happen regardless... and since the `jitter` can return `0` then we can execute without a context switch in the current model (low probability, but possible)

TL;DR - I believe this logic should be purged...

Reply:
push change adding your refactor and removing the delay accounting logic